### PR TITLE
Use tidyselect with get_dupes

### PR DESCRIPTION
**Pull request by jzadra**
Changed the method for selecting variables in the df to use `tidyselect`.  This allows negation, as well as `tidyselect helpers` like `starts_with()`, `ends_with()`, `contains()`, `matches()`, etc.

## Related Issue
#326 

## Example
`mtcars %>% get_dupes(weight = wt, starts_with("cy"))`

## Tests
Includes new tests:

```
test_that("tidyselect specification matches exact specification", {
  expect_equal(mtcars %>% get_dupes(contains("cy"), mpg), mtcars %>% get_dupes(cyl, mpg))
  expect_equal(mtcars %>% get_dupes(mpg), mtcars %>% get_dupes(-c(cyl, disp, hp, drat, wt, qsec, vs, am ,gear, carb)))
  expect_equal(suppressMessages(mtcars %>% select(cyl, wt) %>% get_dupes()), mtcars %>% select(cyl, wt) %>% get_dupes(everything()))
})
```

Also modifies test for specifying zero matching vars, as the old function for doing so was removed.  tidyselect now throws its own error if nothing matches for exact specification.  

NOTE however that using a tidyselect helper that doesn't catch anything will not produce any error or message.

### CODE REVIEW

File: tests/testthat/test-get-dupes.R [hunk 1]
```diff
@@ -37,3 +37,9 @@ test_that("works on variables with irregular names", {
   ) # does it return the right-sized result?
   expect_is(badname_df %>% get_dupes(), "data.frame") # test for success, i.e., produces a data.frame (with 0 rows)
 })
+
+test_that("tidyselect specification matches exact specification", {
+  expect_equal(mtcars %>% get_dupes(contains("cy"), mpg), mtcars %>% get_dupes(cyl, mpg))
+  expect_equal(mtcars %>% get_dupes(mpg), mtcars %>% get_dupes(-c(cyl, disp, hp, drat, wt, qsec, vs, am ,gear, carb)))
+  expect_equal(suppressMessages(mtcars %>% select(cyl, wt) %>% get_dupes()), mtcars %>% select(cyl, wt) %>% get_dupes(everything()))
```
Comment (billdenney) [inline tests/testthat/test-get-dupes.R]: The `suppressMessages()` call suggests that there may be an issue here.  If a message or warning is desired, please use `expect_warning()` or `expect_message()` so that the test includes a test for expected behavior.  If a message or warning is not desired, please update the code so that the warning/message is not generated.

**Comment by billdenney** [comment]
While I've not reviewed the code in detail, the continuous integration check with Travis CI indicates that there is an issue building the janitor.Rmd vignette (https://travis-ci.org/sfirke/janitor/jobs/638671020#L3940).  That suggests that there is some level of backward incompatibility in the change.

Can you please review what is occurring in the vignette and do the following:

* Add a test for the old functionality as occurred in the vignette
* Make sure that the new code works like the old code did (i.e. preserve backward compatibility)

**Comment by billdenney** [comment]
I made a comment on your last test that I'm not sure if it will show up because there is a "pending" message above it.  If you get this twice, sorry, this is to ensure that you get it at least once:

The `suppressMessages()` call suggests that there may be an issue here.  If a message or warning is desired, please use `expect_warning()` or `expect_message()` so that the test includes a test for expected behavior.  If a message or warning is not desired, please update the code so that the warning/message is not generated.

File: tests/testthat/test-get-dupes.R (hunk 1 above)
Comment (jzadra) [reply tests/testthat/test-get-dupes.R]: The message from an empty `get_dupes()` call is a custom message that says `"No variable names specified - using all columns.\n"` (line 28).  This message doesn't occur when using `everything()` because you have specified variable names. My test was intended to compare the outputs of the two calls while ignoring that one had that custom message.  What's a better way to do this?

**Comment by jzadra** [comment]
The only error I get when running the janitor.Rmd vignette is 

> While I've not reviewed the code in detail, the continuous integration check with Travis CI indicates that there is an issue building the janitor.Rmd vignette (https://travis-ci.org/sfirke/janitor/jobs/638671020#L3940). That suggests that there is some level of backward incompatibility in the change.
> 
> Can you please review what is occurring in the vignette and do the following:
> 
> * Add a test for the old functionality as occurred in the vignette
> * Make sure that the new code works like the old code did (i.e. preserve backward compatibility)

It looks like it was just that I forgot to add tidyselect to the imports in DESCRIPTION.  I've made the change on my fork, how do I add them to this PR?

**Comment by billdenney** [comment]
There still appears to be an issue with the `janitor.Rmd` vignette.  Can you check that one?  Also, you can do most of the checks that will happen on Travis by running `devtools::check()`.

**Comment by jzadra** [comment]
I've run devtools::check().  The only error I get is:

```
── 1. Failure: if called on non-existent vector, returns use
`mtcars %>% tabyl(moose)` threw an error with unexpected message.
Expected match: "object 'moose' not found"
Actual message: ""
```

**Comment by sfirke** [comment]
Oh!  I think that error is caused by the newer tidyselect, this is why I have to resubmit to CRAN.  Let me see about patching that test right now.

**Comment by sfirke** [comment]
I just committed this to master but try implementing it also in your branch: change line 190 of `test-tabyl.R` to be `  expect_error(mtcars %>% tabyl(moose))`

**Comment by jzadra** [comment]
Yes that solved the moose error.  Interesting, I also had to change line 29 of test_get_dupes.R in the same way (removing the message).

**Comment by jzadra** [comment]
It's still not finding tidyselect on travis.

**Comment by sfirke** [comment]
The failing error message is:

> Quitting from lines 103-104 (janitor.Rmd) 
> 
> Error: processing vignette 'janitor.Rmd' failed with diagnostics:
> 
> 'eval_select' is not an exported object from 'namespace:tidyselect'
> 
> --- failed re-building ‘janitor.Rmd’

Looks like `eval_select` is only present in the dev version of tidyselect on GitHub?  There's a hit for it here: https://github.com/r-lib/tidyselect/search?q=eval_select&unscoped_q=eval_select and not here https://github.com/cran/tidyselect/search?q=eval_select&unscoped_q=eval_select (the current CRAN version).

Are you using a function that's only in the new version of tidyselect?  Ideally we wouldn't need to require the newest version of tidyselect, that would complicate getting both packages to CRAN.  If the same functionality could be implemented with existing tidyselect functions that aren't being made obsolete, that would be the best.

**Comment by sfirke** [comment]
If you must use functions from upcoming tidyselect 1.0, or feel that they're superior and the way to go, it wouldn't be too bad to make janitor require tidyselect 1.0.  It doesn't have so many breaking changes that I think it would bother people to make them update it.

To do this with CRAN, I could submit a simple patch to handle the change in error message, then submit this feature as part of a larger 1.4 release following tidyselect 1.0 going to CRAN.

**Comment by jzadra** [comment]
Looks like adding r-lib/tidyselect to the remotes has worked.  Would you prefer to wait until tidyselect 1.0.0 is on cran and set that version dependency in Imports?

File: tests/testthat/test-get-dupes.R (hunk 1 above)
Comment (sfirke) [reply tests/testthat/test-get-dupes.R]: You're good here then, thanks for explaining.

File: DESCRIPTION [hunk 2]
```diff
@@ -26,7 +26,8 @@ Imports:
     snakecase (>= 0.9.2),
     magrittr,
     purrr,
-    rlang
+    rlang,
+    tidyselect
```
Comment (sfirke) [inline DESCRIPTION]: let's specify at least version 1.0.0 now that it's on CRAN

File: DESCRIPTION [hunk 3]
```diff
@@ -36,5 +37,7 @@ Suggests:
     rmarkdown,
     sf,
     tibble
+remotes:
+    r-lib/tidyselect
 VignetteBuilder: knitr
```
Comment (sfirke) [inline DESCRIPTION]: and I think we can remove this now that 1.0.0 is on CRAN

File: R/get_dupes.R [hunk 4]
```diff
@@ -3,47 +3,48 @@
...
   } else {
-    nms <- rlang::quos(...) %>%
-      rlang::quos_auto_name()
-    var_names <- names(nms)
+    var_names <- names(pos)
+    nms <- rlang::syms(var_names)
   }
-
-  # check that each variable name provided is present in names(dat); if not, throw error
-  check_vars_in_df(dat, df_name, var_names)
```
Comment (sfirke) [inline R/get_dupes.R]: is this removed b/c it was used to deal with how the old tidyverse functions errored and now they error in the desired way already -- i.e., they identify the mis-specified variable -- so it's now redundant?

**Comment by jzadra** [comment]
Hi, sorry for the delay. I'm not sure how to integrate your changes with my local fork that is linked to the PR.  Is there a way to pull the changes you've made back into my local fork, or do I need to just manually copy paste from here into my fork? Some other method?  Sorry, first time!

File: R/get_dupes.R (hunk 4 above)
Comment (jzadra) [reply R/get_dupes.R]: Correct.  Tidyselect will throw its own error if an exact column specification doesn't match the table.

But just a reminder (we'e discussed previously) that any matching functions (starts_with, etc) that don't catch any variables will not generate any error currently.

File: DESCRIPTION (hunk 2 above)
Comment (jzadra) [reply DESCRIPTION]: done.  And thanks for the invite to add myself as contrib!

File: DESCRIPTION (hunk 3 above)
Comment (jzadra) [reply DESCRIPTION]: done

**Comment by jzadra** [comment]
Awesome, thanks for the coaching through my first PR and contribution!
